### PR TITLE
[omni, model, config, docs] feat: OmniConfig / OmniModel / graph FSM / processing / mixins

### PR DESCRIPTION
**Pull request by OnePunchMonk**
## Summary

Part of #1057 (SeedOmni V2 upstreaming plan) — implements #1062: the V2 core primitives, with no concrete model attached, so the graph/config/processing contracts get reviewed on their own terms before the Wave 3 model PRs start depending on them.

Depends on: #1058 (merged), #1060 (merged via #1090).

## What's in scope here

- `configuration_omni.py` / `modeling_omni.py` / `omni_pretrained_model.py` — `OmniConfig` (a plain HF `PretrainedConfig`), the eager graph forward/generate `OmniModel`, and the base `PreTrainedModel` wrapper.
- `graphs/` — training + generation FSM definitions (node/edge/end semantics), validation, mermaid export.
- `processing/` + `processing_omni.py` — `OmniProcessor.from_config()` composing the active modules' CPU preprocessors.
- `mixins/` — `base_mixin`, `training_module_mixin`, `inference_module_mixin`, `metric_meter_mixin`, `offline_encoding_mixin`, `emb_parallel_mixin`.
- `utils/` — `OmniModuleCheckpointManager`, conversation carrier, graph profiler, visualize, offline cache, convert registry, and a new `model_path.py::resolve_model_path` helper.
- `modules/` — `OMNI_{CONFIG,MODEL,PROCESSOR}_REGISTRY` plus the shared `modules/base/` (`text_encoder/` + `llm_packing.py`), which later Wave-3 model PRs (bagel, janus, qwen3, qwen3_moe, qwen3vl) register into without touching this PR's code.
- `loader.py` — consults the OMNI registries so `build_foundation_model` / `build_processor` work for an OmniModule sub-module exactly like any other model.

## What's deliberately out of scope

- `accelerator/` (`ModuleRuntime`, `OmniModelRuntime`, `execute_train_node`/`execute_generation_node`) — the FSDP-aware runtime layer is #1061/#1064.
- `OmniTrainer` / `OmniInferencer` / `OmniArguments` (the launcher argument schema) — #1064.
- The SeedOmni V2 data pipeline (`SeedOmniCollator`, conversation transform) — #1063.

The `omniv2` draft branch (#1056) doesn't actually split cleanly along these lines yet, so a few of the test files nominally listed under #1062's CI section needed adjustment against this PR's narrower scope:

- `test_graph.py` and `test_offline_encoding_mixin.py`: kept, with the handful of cases that exercised `OmniModelRuntime`/`execute_train_node`/`OmniTrainer`/`GraphProfileCallback` (i.e. #1064) or a concrete `bagel` module removed. Everything that exercises pure graph FSM / `OmniModel` / mixin behavior is kept.
- `test_omni_config_paths.py`: kept, but its target (`_resolve_model_path`) is extracted standalone into `seed_omni/utils/model_path.py` rather than pulled in via the full ~1000-line `OmniArguments` schema (#1064's file).
- `test_modules_mixin.py` and `test_preprocessor.py`: dropped from this PR — the large majority of both files' cases directly exercise `bagel`/`janus`/`qwen3`/`qwen3vl` concrete modules that don't exist until their respective Wave-3 PRs land.
- `test_img_tag.py` and `test_offline_cache_writer.py`: dropped — both test the SeedOmni V2 data pipeline (`veomni.data.seed_omni.*`), which is #1063's file scope, not #1062's.

Happy to move any of these back in, or split further, if a maintainer would rather see them land differently.

## CI

Wires `tests/seed_omni/` into `.github/workflows/gpu_unit_tests.yml` and `.github/workflows/npu_unit_tests.yml` — all CPU-only, no GPU/NPU-specific code touched.

## Test plan

```
pytest -q tests/seed_omni/
```
73 passed. Ran in a clean Python 3.12 container on Modal (torch 2.14.0, transformers 5.16.1, plus the repo's non-GPU base dependencies — the `gpu` extra's CUDA-pinned wheels weren't needed since nothing in this PR touches CUDA/Triton kernels).

`ruff check` and `ruff format --check` both clean on all new/changed files.

## Acceptance checklist (from #1062)

- [x] `pytest tests/seed_omni/` green on CPU
- [x] `OmniConfig` round-trips through `save_pretrained` / `from_pretrained` — independently verified beyond the test suite with a standalone script (multi-module config, `infer_type`/`generation_kwargs` set, checked all public accessors after reload, plus a second save→load pass for stability). `save_pretrained` writes `config.json` + `training_graph.yaml` + `generation_graph.yaml` + a `graphs/` dir (not a single flat JSON), and all of it round-tripped correctly.
- [x] Graph validation rejects dangling nodes, unknown endpoints and unreachable ends with actionable errors (see `test_graph.py`)
- [x] No import of `veomni.arguments` from `configuration_omni.py`

## Summary by CodeRabbit

- **New Features**
  - Added SeedOmni V2 support for composing, loading, generating with, and saving multimodal models.
  - Added conversation processing for text and images, including tokenization, chat templates, and batched requests.
  - Added configurable training and generation graphs with Mermaid visualization.
  - Added checkpoint conversion, per-module checkpoint management, offline caching, and model asset handling.
  - Added text encoder support with sampling, parallel execution, and metric collection.

- **Tests**
  - Added comprehensive coverage for conversations, graphs, processors, checkpoint handling, configuration, and offline encoding.
  - Added SeedOmni V2 test suites to GPU and NPU workflows.

### CODE REVIEW

**Comment by OnePunchMonk** [comment]
## Test environment + results (for reviewer reference)

Both runs below were executed in a fresh Modal container (not local, not CI) to get a clean-room result against this PR's actual diff.

### Environment

- Base image: `debian_slim`, Python 3.12
- Hardware: CPU-only container — no GPU attached (nothing in this PR touches CUDA/Triton kernels, so none was needed)
- apt: `git`, `libgl1`, `libglib2.0-0`
- pip (key versions): `torch==2.14.0`, `transformers==5.16.1`, `pytest==7.4.4`, `datasets==2.21.0`, `pyarrow==25.0.1`, `numpy==2.5.3`, `av==18.1.0`, `decord==0.6.0`, plus the repo's other non-`gpu`-extra base deps (`psutil`, `timm`, `wandb`, `einops`, `matplotlib`, `blobfile`, `tiktoken`, `torchdata`)
- Repo mounted at `/repo` from this branch's exact working tree (no `pip install -e .`, imported directly via `sys.path`)

### Run 1 — full scoped test suite

```
pytest -q tests/seed_omni/
```

Result: **73 passed** in 14.97s.

Modal run: https://modal.com/apps/onepunchmonk/main/ap-nkm6PQ3SFJNNlbklLPcPBf

Also ran (same container/image, not on Modal GPU, plain lint):
- `ruff check` — all checks passed
- `ruff format --check` — 45 files already formatted

### Run 2 — independent `OmniConfig` round-trip verification

Standalone script (not part of the pytest suite) exercising `save_pretrained` → `from_pretrained` directly, since the acceptance criterion asks for this explicitly:

- Built an `OmniConfig` with 2 modules, a training graph, a generation graph, `infer_type="infer_gen"`, `generation_kwargs={"max_new_tokens": 128}`
- `save_pretrained(tmp_dir)` → confirmed on-disk layout: `config.json`, `training_graph.yaml`, `generation_graph.yaml`, `graphs/` (not a single flat JSON)
- `from_pretrained(tmp_dir)` → checked all 11 public accessors match the original: `modules`, `training_graph`, `generation_graphs`, `infer_type`, `generation_kwargs`, `module_names`, `training_edges`, `infer_types`, `generation_graph`, instance type, `model_type`
- Re-saved the *reloaded* config and reloaded again (double round-trip) → `to_dict()` identical, confirming stability under repeated cycles

Result: **ALL_OK** (12/12 checks, including the double round-trip check).

Modal run: https://modal.com/apps/onepunchmonk/main/ap-mn0FGSQ1FoDxvzY0rAe1CI

### Notes for review

- Neither run used the `gpu` extra (`torch==2.11.0+cu130`, flash-attn, etc.) from `pyproject.toml` — deliberately, since none of this PR's code touches GPU-specific kernels. If a maintainer wants the exact CI dependency graph (`uv sync --frozen --extra gpu --group dev`) re-verified on real GPU hardware, that's a separate, heavier run I can do on request.
- Full package list and container build log are in the Modal run links above if you want to check for anything I didn't call out here.

**Comment by OnePunchMonk** [comment]
@Coach257 saw that this pr's scope matches #1192 and looks to supersede it. If there's other unclaimed work in #1057 that wouldn't overlap with your planned picks, will be glad to pick that up next. Also, could you take a look at #1185 when you get a chance ?

**Comment by Coach257** [comment]
Thanks! I will check the difference between this pr and #1192. OmniModel and related should be built based on modelruntime #1125 @OnePunchMonk .

**Comment by Coach257** [comment]
Thanks for this, and for the scoping write-up in particular — it was genuinely useful to read against our own split of #1062.

On how this will land: the V2 core is going in from our internal `omniv2` branch, so we'll take that as the source of truth here and close this PR, along with the others you opened against the #1057 series. That's not a judgement on the work — the blocker is that this branch snapshots `omniv2` from before 2026-09-03, and the core has moved since. Two concrete examples: `OmniModel.forward` here still raises `NotImplementedError` and defers training to the runtime, which we replaced on 09-03 with a `node_runner` seam that keeps the training-graph walk in the core; and `modules/base/text_encoder/accelerated.py` became an `accelerated/` subpackage on 09-08. Rebasing across those is more work than landing the current tree.

We are keeping this PR as a reference and cross-checking our slice against it. One thing that comparison surfaced, in case it's useful for future work in this area: several of the core's dependencies live *outside* `veomni/models/seed_omni/` — `veomni/ops/kernels/embed/` (for `emb_parallel_mixin`), `veomni/utils/tensor_utils.py` (for `llm_packing` and the accelerated text encoder), and the `loss_reduction_group` parameter on `ForCausalLMLoss` — so a selection made by path prefix leaves the tree importable only as long as nothing actually resolves the lazy registry factories.

Thanks again for the effort here.
